text/0000-structural-records.md

```diff
@@ -0,0 +1,1298 @@
+they are written in and do not have a specified type layout according to that
+non-existent order. This means that they would not be particularly useful for
+FFI purposes. In 2016, [@retep998] noted that they would like to use such
+records to *"closely match what the headers are doing"*. @retep988 later
+noted this in a later [comment][@retep998_2].
+
+In the future however, to support that use case, we could allow attributes
+on types which would make `#[repr(C)] { x: A, y: B }` legal to write.
+In that case, the order would be respected and so it could be used for FFI.
+This would entail that `#[repr(C)] { x: A, y: B }` would be a different
+type than `{ x: A, y: B }`.
+
+## `HasField<"the_field">`
+
+Using const generics, the compiler could automatically implement a trait
+`Get<"the_field">` or `HasField<"the_field">` and the corresponding setter
+for any type which has that field. This would offer a way to be polymorphic
+over fields. The `HasField` trait could be defined as:
+
+```rust
+#[lang_item = "has_field"]
+trait HasField<const Field: &'static str> {
+    type Field;
+
+    fn get(self) -> Self::Field;
+
+    fn get_ref<'a>(&'a self) -> &'a Self::Field;
+
+    fn get_mut<'a>(&'a mut self) -> &'a mut Self::Field;
+}
+```
+
+Other designs may involve generic associated types (GATs) or some
+form of polymorphism over type constructors (i.e. higher kinded types).
+
+[RFC 2529]: https://github.com/rust-lang/rfcs/pull/2529
+
+To take privacy into account, hidden implementations could be used as
+in [RFC 2529].
+
+[OverloadedRecordFields]: https://github.com/ghc-proposals/ghc-proposals/pull/6
+
+This idea mainly comes from Haskell and is seen in the [SuperRecord]
+package aforementioned. See [OverloadedRecordFields] for a deeper
+discussion.
+
+## Supercharging FRU
+
+The current FRU (struct update) syntax is rather limited.
+For example, you cannot write `MyType { ..x, ..y }` to get some fields
+from both `x` and `y`. This is understandable, after all, since both `x`
+and `y` are fully formed `MyType`s, this would just take all the values
+from `x` without taking anything from `y`.
+
+However, for structural records, this wouldn't be the case.
+You could imagine merging two structural records together with:
+
+```rust
+let alpha = { a: 1, b: 2 };
+let beta  = { c: 3, d: 4 };
+let gamma = { ..alpha, ..beta };
+
+assert_eq!(gamma, { a: 1, d: 2, c: 3, d: 4 });
+```
+
+## Coercions from larger to smaller records
+
+The syntax above allowed us to grow structural records with more fields.
+We could also offer a way to shrink them by allowing coercions from larger
+to smaller structs. This would partially move (or copy) all the bits you
+used. For example:
+
+```rust
+type Smaller = { foo: u8, bar: bool };
+type Bigger = { foo: u8, bar: bool, baz: char };
+
+let b: Bigger = { foo: 1, bar: true, baz: 'a' };
+let s: Smaller = b; // OK! `b` has all the fields `Smaller` needs.
```

Comment:
This causes similar problems as with subtyping or implicit conversion in general. A more explicit solution is recommend, for instance with the as operator:
`let s: b as Smaller`

The same with nominal<->structural conversions.

Reply:
Not exactly; implicit conversions have problems (mainly in terms of giving type inference too many options to pick from sometimes as well as not being as clear about semantics), but they are nothing like subtyping which can affect and constrain memory layout / representation.

The main difference between subtyping and implicit coercion is that the latter only affects top level constructions while the former (when covariance occurs) makes it such that:
```rust
Vec<{ alpha: T, }> <: Vec<{ alpha: T, beta: U }>
```
For a systems language this would be quite problematic.

That said, because implicit conversions do have problems, I have left it out of the main proposal at the moment.

Reply:
>The main difference between subtyping and implicit coercion is that the latter only affects top level constructions while the former (when covariance occurs) makes it such that:

Hmm..., so you would exclude covariance of structuralism.